torch/csrc/jit/runtime/static/ops.cpp

```diff
@@ -670,6 +674,83 @@ REGISTER_OPERATOR_FUNCTOR(aten::nan_to_num, aten_nan_to_num, [](Node* n) -> SROp
   };
 });
 
+namespace {
+
+class VarStackNodeWrapper {
+ public:
+  explicit VarStackNodeWrapper(const ProcessedNode& pnode) : pnode_(pnode) {}
+
+  const at::Tensor& operator[](size_t idx) const {
+    TORCH_CHECK(idx < size());
+    return pnode_.Input(idx).toTensor();
+  }
+
+  size_t size() const {
+    return pnode_.num_inputs() - 1;
+  }
+
+ private:
+  const ProcessedNode& pnode_;
+};
+
+void varStackSerialOut(
+    at::Tensor& result,
+    int64_t dim,
+    const VarStackNodeWrapper& inputs) {
+  auto result_sizes = inputs[0].sizes().vec();
+  result_sizes.insert(result_sizes.begin() + dim, inputs.size());
+  at::native::resize_(result, result_sizes);
+
+  AT_DISPATCH_FLOATING_TYPES(
+      result.scalar_type(), "varstack_serial_kernel", [&]() {
+        at::native::detail::
+            stack_serial_kernel_impl<scalar_t, VarStackNodeWrapper>(
+                result, inputs, dim);
+      });
+}
+
+std::vector<at::Tensor> unsqueezeVarStackInputs(
+    const VarStackNodeWrapper& inputs,
+    const int64_t dim) {
+  std::vector<at::Tensor> result;
+  result.reserve(inputs.size());
+  for (const auto i : c10::irange(inputs.size())) {
+    result.push_back(at::native::unsqueeze(inputs[i], dim));
+  }
+  return result;
+}
+
+void varstackNonserialOut(
+    at::Tensor& result,
+    const int64_t dim,
+    const VarStackNodeWrapper& inputs) {
+  std::vector<at::Tensor> inputs_unsqueezed =
+      unsqueezeVarStackInputs(inputs, dim);
+  fastResizeToZero(result);
+  at::native::_cat_out_cpu(inputs_unsqueezed, dim, result);
+}
+
+void varStackOut(ProcessedNode& pnode, int64_t dim) {
```

Comment:
Hey @mikeiovine! It looks like this PR adds a custom `stack` kernel for static runtime to avoid the overhead from refcount bumps, which is incurred from converting from your boxed representation (`ProcessedNode`) to the unboxed representation (`ITensorList` <-> `ArrayRef<Tensor>`).

In core we're working on a way to improve this situation (see the [github issue](https://github.com/pytorch/pytorch/issues/66328) for it), and @ysiraichi has a [stack of PR's](https://github.com/pytorch/pytorch/pull/67964#discussion_r756689701) out.

The high level idea is to replace the `TensorList` argument to `stack` (and all other ops that take lists of tensors as arguments) with a new, read-only `ITensorList` type, which under the hood can be easily created from unboxed types (`ArrayRef<Tensor>`) or boxed types (`c10::List<Tensor>`), and doesn't incur refcount bumps.

This PR required adding a custom `stack` kernel for static runtime. I think we should be able to remove most of the logic here (and the changes in `CanUseNativeSerialStack`)  if we also teach the `ITensorList` type how to be convertible from `ProcessedNode`. That should also give static runtime similar perf improvements for every other operator that we eventually update to use `ITensorList` arguments.

Let me know if that sounds good to you!

Reply:
Sure, that sounds good to me. FYI I currently have [a PR out which generalizes the idea from this one](  https://github.com/pytorch/pytorch/pull/71573). Maybe we should make `ITensorList` convertible from `ProcessedNodeInputWrapper`? It only allows read-only access to node inputs and ensures that they are actually tensors (technically `ProcessedNode` can hold any kind of `IValue`)

Reply:
Yeah, I think that would be fine - or if `ITensorList` could be convertible directly from a `ProcessedNode`, then it would also directly have to do the checks to ensure that the Processed Node actually contains a list of tensors.

Reply:
The other nice thing about `ProcessedNodeInputWrapper` is that it allows you restrict access to a subset of the node inputs - sometimes, we're only interested in the first N. For example, in `VarStack`, the last input needs to be ignored (it's the stack `dim`)